Implement URL patterns for webengine cookies.accept

lufte opened the pull request:
This PR is a sub-task of issue #3636 

TODO: tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4395)
<!-- Reviewable:end -->

lufte commented:
Question: do we want QtWebkit support for this? I've found that it's not as trivial as this one, but I don't know how you feel about supporting url patterns only for one backend.

The-Compiler commented:
Hmm, hard to say. I'd say we either need support for both backends, or need support for something like `supports_pattern: QtWebEngine`. I'd actually prefer the latter, as I guess we might need it elsewhere as well (and I'd rather not spend time adding new QtWebKit features).

lufte commented:
Yeah, something like `supports_pattern: QtWebEngine` would indeed be useful.

I think it's pretty easy to support `never` and `all` with url patterns in QtWebkit, a couple of `if`s in `browser.webkit.cookies.RAMCookieJar.setCookiesFromUrl` would do it. But the logic for `no-3rdparty` and `no-unknown-3rdparty` seems to be in another Qt class that qutebrowser is not currently subclassing, and we don't have enough information in `RAMCookieJar` to do it there.

The-Compiler commented:
Note that `RAMCookieJar` is used for private browsing only. For normal browsing, the logic in `qutebrowser/browser/webkit/webkitsettings.py` is relevant, which would need to be done on a per-domain basis instead. Others are handled in `AbstractSettings.update_for_url` in `qutebrowser/config/websettings.py`, but since this is only relevant for QtWebKit, it gets a bit more tricky. Probably not worth the effort.

lufte commented:
Right, but `CookieJar` inherits from `RAMCookieJar`, specially that method where the check for `never` is made. I've tested it with a non-private window and it definitely kicks in.

I haven't seen that `AbstractSettings.update_for_url` method yet, maybe it's not that hard to implement after all. I'll keep looking just in case, though I agree that `supports_pattern: QtWebEngine` is needed nevertheless.

Thanks for your comments, have a nice day.

jgkamat commented:
I don't know if we need `supports_pattern: QtWebEngine`. If we did, it would cause anyone who uses this feature to have a broken config on starting with webkit. I think silently ignoring the url patterns on webkit would be fine, personally

The-Compiler commented:
FWIW I'm fine with ignoring the URL patterns on QtWebKit. I'd also be fine with merging this without tests, seeing that our current testbase for per-domain settings is kind of meager anyways.

However, one thing I'm not sure about: Should this check for `request.origin` ("The URL of the script or content accessing a cookie. Can be used to white-list or black-list third-party cookie access for specific services.") or `request.firstPartyUrl` ("The URL that was navigated to. The site that would be showing in the location bar if the application has one. Can be used to white-list or black-list cookie access or third-party cookie access for specific sites visited.")? ([docs](https://doc.qt.io/qt-5/qwebenginecookiestore-filterrequest.html))

I think from how other settings work right now, `firstPartyUrl` would be the more logical choice, no? Using `origin` is then what something uMatrix-like would probably do?

toofar commented:
I wrote a [while ago](https://github.com/qutebrowser/qutebrowser/issues/3636#issuecomment-413135035) about how `origin` had a lot more utility. But I think at this point we have enough per-url settings with the same semantics that it wouldn't make sense to make one of them regarding the URLs being requested rather than the URL that you were navigated to.

lufte commented:
Hi, sorry for the late response. I think that the `no-3rdparty` value becomes kind of confusing if we use `origin` instead of `firstPartyUrl`. When I set `content.cookies.accept` to `no-3rdparty` for `domain.com`, what I expect is that no third party cookies are accepted when visitng `domain.com`, not that "no cookies from `domain.com` are accepted while visiting other sites".

I agree though that the second option is as (or even more) useful, though. Maybe new values are needed for `content.cookies.accept`, or a whole new setting. 

Anyway, for now I changed the property and added a test. I didn't have time to run the build locally so I'll leave it running here and fix whatever breaks later.

lufte commented:
It seems pytest complains about two test files having the same name. They used to recommend not having `__init__.py` files in test directories (which would fix this) but they don't anymore. Are you still following that recommendation or do I add the files?

https://stackoverflow.com/questions/12582503/py-test-test-discovery-failure-when-tests-in-different-directories-are-called#12600278

jgkamat commented:
Javier Ayres writes:

> It seems pytest complains about two test files having the same name. They 
> used to recommend not having `__init__.py` files in test directories (which 
> would fix this) but they don't anymore. Are you still following that 
> recommendation or do I add the files?

I would just rename your test file to unblock yourself. Maybe 
test_webengine_cookies.

As for review comments, could you add a note to the setting docstring saying 
that the per-domain functionality will be ignored on webkit.

jgkamat commented on tests/unit/browser/webengine/test_cookies.py:
```diff
@@ -0,0 +1,54 @@
+    config_stub.set_str('content.cookies.accept', 'all')
+    config_stub.set_str('content.cookies.accept', 'no-3rdparty',
+                        pattern=UrlPattern('https://*.domain1.com'))
+    assert not _accept_cookie(request)
```
Could you add at least one test of a request that does not hit the per-url setting? Just for completeness :)

lufte commented on tests/unit/browser/webengine/test_cookies.py:
```diff
@@ -0,0 +1,54 @@
+    """
+    request = MagicMock()
+    request.firstPartyUrl = QUrl('https://domain1.com')
+    request.thirdParth = False
```
Oops. Yes.

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
```
nitpick: Please add an empty line between the comment and imports.

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2015-2019 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```
```suggestion
# Copyright 2019 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+from PyQt5.QtCore import QUrl
+from unittest.mock import MagicMock
+from qutebrowser.browser.webengine.cookies import _accept_cookie
+from qutebrowser.utils.urlmatch import UrlPattern
```
We try to always import modules rather than things from modules (except for Qt things which are clearly namespaced via the Q prefix). That way, it's clearly visible where something is coming from, which IMHO makes things more readable.

In other words:

```python
from qutebrowser.utils import urlmatch
from qutebrowser.browser.webengine import cookies
```

etc.

Also, please structure imports as per [pep8](https://www.python.org/dev/peps/pep-0008/#imports):

```
stdlib imports

third-party imports

qutebrowser imports
```

(with a blank line in between)

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+
+def test_accept_cookie(config_stub):
+    """Test that _accept_cookie respects content.cookies.accept."""
+    request = MagicMock()
```
Why use a mock at all? Like you've seen, that allows getting attributes which were never defined, which can be troublesome.

Instead I'd use the real thing, but move it to a [pytest fixture](https://docs.pytest.org/en/latest/fixture.html) so the tests are skipped on older Qt versions where cookie filters aren't available. Something like this (untested):

```python
import pytest
from PyQt5.QtWebEngineCore import QWebEngineCookieStore

@pytest.fixture
def filter_request():
    try:
        return QWebEngineCookieStore.FilterRequest()
    except AttributeError:
        pytest.skip("FilterRequest not available")
```

Then you can just add a `filter_request` argument to tests which need one.

(Note: Don't call it `request` because that's [a special pytest fixture](https://docs.pytest.org/en/latest/reference.html#request))

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+    """Test that _accept_cookie respects content.cookies.accept."""
+    request = MagicMock()
+
+    config_stub.set_str('content.cookies.accept', 'all')
```
For consistency with other tests, use `config_stub.val.content.cookies.accept = 'all'` to set settings.

lufte replied:
This only works when not using url patterns, right? Otherwise I don't know how to set the setting like this with a pattern.

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+    assert _accept_cookie(request)
+
+    config_stub.set_str('content.cookies.accept', 'never')
+    assert not _accept_cookie(request)
```
Please use [pytest parametrization](https://docs.pytest.org/en/latest/parametrize.html) to split this into individual subtests. Something like (untested):

```python3
import pytest

@pytest.mark.parametrize('value, third_party, accepted', [
    ('never', False, False),
    # ...
])
def test_accept_cookie(config_stub, filter_request, value, third_party, accepted):
    config_stub.val.content.cookies.accept = value
    filter_request.thirdParty = third_party
    assert cookies._accept_cookie(filter_request) == accepted
```

The-Compiler commented on tests/unit/browser/webengine/test_webengine_cookies.py:
```diff
@@ -0,0 +1,73 @@
+    assert not _accept_cookie(request)
+
+
+def test_accept_cookie_with_pattern(config_stub):
```
Things from above (real `FilterRequest` and parametrization) probably also apply here.

As for parametrizing, I guess you could parametrize `value, pattern, pattern_value` or so.

lufte commented:
```
tests/unit/browser/webkit/test_tabhistory.py ........................F
_____________________________ test_titles[1-item1] _____________________________
objects = Objects(history=<PyQt5.QtWebKit.QWebHistory object at 0x7fbf99d70588>, user_data=[None, None, {'foo': 23, 'bar': 42}, {'zoom': 149, 'scroll-pos': PyQt5.QtCore.QPoint()}, {'zoom': 120, 'scroll-pos': PyQt5.QtCore.QPoint()}])
i = 1
item = qutebrowser.misc.sessions.TabHistoryItem(active=True, original_url=PyQt5.QtCore.QUrl('http://example.com/…'), title='percent', url=PyQt5.QtCore.QUrl('http://example.com/…'), user_data=None)
    @pytest.mark.parametrize('i, item', enumerate(ITEMS))
    def test_titles(objects, i, item):
        """Check if the titles were loaded correctly."""
>       assert objects.history.itemAt(i).title() == item.title
E       AssertionError: assert 'Network access is disabled' == 'percent'
E         - Network access is disabled
E         + percent
```
This doesn't look to be related to my changes. @jgkamat do you know what this is about?

jgkamat commented:
Javier Ayres writes:

> ```
> tests/unit/browser/webkit/test_tabhistory.py ........................F
> _____________________________ test_titles[1-item1] 
> _____________________________
> objects = Objects(history=<PyQt5.QtWebKit.QWebHistory object at 
> 0x7fbf99d70588>, user_data=[None, None, {'foo': 23, 'bar': 42}, {'zoom': 
> 149, 'scroll-pos': PyQt5.QtCore.QPoint()}, {'zoom': 120, 'scroll-pos': 
> PyQt5.QtCore.QPoint()}])
> i = 1
> item = qutebrowser.misc.sessions.TabHistoryItem(active=True, 
> original_url=PyQt5.QtCore.QUrl('http://example.com/…'), title='percent', 
> url=PyQt5.QtCore.QUrl('http://example.com/…'), user_data=None)
>     @pytest.mark.parametrize('i, item', enumerate(ITEMS))
>     def test_titles(objects, i, item):
>         """Check if the titles were loaded correctly."""
>>       assert objects.history.itemAt(i).title() == item.title
> E       AssertionError: assert 'Network access is disabled' == 'percent'
> E         - Network access is disabled
> E         + percent
> ```
> This doesn't look to be related to my changes. @jgkamat do you know what 
> this is about?

Looks like a bad test, it's failing on master too. I would just ignore it.

Earnestly commented:
I've been using this locally with some success.  In addition this would it be possible for per-domain `content.cookies.store` as well?  In this manner I can globally block and not store all cookies except for a few whitelisted domains.

Example usage:
```py
try:
    # E.g.
    #   https://github.com
    #   https://ebay.co.uk all

    whitelist = [(pair[0], pair[1] if len(pair) > 1 else 'no-3rdparty')
                 for pair in (line.split(' ')
                 for line in open(config.configdir / 'cookie-whitelist').read().splitlines())]

    for domain, license in whitelist:
        config.set('content.cookies.accept', license, domain)
        config.set('content.cookies.store', True, domain)
except FileNotFoundError:
    pass
```

The-Compiler commented:
@Earnestly Thanks for the feedback! Will look at getting this merged soon, as it seems to look fine, and people testing things is always a big help.

It's not possible to set `content.cookies.store` per-domain (as QtWebEngine only has that option exposed per storage profile), but in your case, I don't see the necessity: If you don't accept cookies from a given domain, there's nothing to store anyways - so setting `content.cookies.store` to `True` globally will have the same effect as your snippet.

Earnestly commented:
That's a fair point.

I'd like to add though that I am getting some odd behaviour.  With that script snippet I set `ebay` to use `all` for it's accept, but it doesn't appear to be working.  I have to manually set `cookie.accept` to `all` and reload the page to restore some missing functionality (in this case, the ability to view larger pictures).

The-Compiler commented:
Just got this from @Earnestly:

```
15:41:50 ERROR: Uncaught exception
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/qutebrowser/browser/webengine/cookies.py", line 28, in _accept_cookie
    accept = config.instance.get('content.cookies.accept',
  File "/usr/lib/python3.8/site-packages/qutebrowser/config/config.py", line 365, in get
    obj = self.get_obj(name, url=url, fallback=fallback)
  File "/usr/lib/python3.8/site-packages/qutebrowser/config/config.py", line 389, in get_obj
    value = self._values[name].get_for_url(url, fallback=fallback)
  File "/usr/lib/python3.8/site-packages/qutebrowser/config/configutils.py", line 236, in get_for_url
    if scoped.pattern is not None and scoped.pattern.matches(url):
  File "/usr/lib/python3.8/site-packages/qutebrowser/utils/urlmatch.py", line 292, in matches
    qtutils.ensure_valid(qurl)
  File "/usr/lib/python3.8/site-packages/qutebrowser/utils/qtutils.py", line 158, in ensure_valid
    raise QtValueError(obj)
qutebrowser.utils.qtutils.QtValueError: PyQt5.QtCore.QUrl('') is not valid
15:41:50 WARNING: QSqlDatabasePrivate::database: requested database does not belong to the calling thread.
15:41:50 WARNING: QSqlDatabasePrivate::database: requested database does not belong to the calling thread.
15:41:50 WARNING: QSqlQuery::prepare: database not open
15:41:50 ERROR: Error while getting objects
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/qutebrowser/misc/crashsignal.py", line 212, in _get_exception_info
    all_objects = debug.get_all_objects()
  File "/usr/lib/python3.8/site-packages/qutebrowser/utils/debug.py", line 364, in get_all_objects
    _get_pyqt_objects(pyqt_lines, start_obj)
  File "/usr/lib/python3.8/site-packages/qutebrowser/utils/debug.py", line 347, in _get_pyqt_objects
    lines.append('    ' * depth + repr(kid))
  File "/usr/lib/python3.8/site-packages/qutebrowser/browser/history.py", line 189, in __repr__
    return utils.get_repr(self, length=len(self))
  File "/usr/lib/python3.8/site-packages/qutebrowser/misc/sql.py", line 324, in __len__
    q = Query("SELECT count(*) FROM {table}".format(table=self._name))
  File "/usr/lib/python3.8/site-packages/qutebrowser/misc/sql.py", line 185, in __init__
    self._check_ok('prepare', ok)
  File "/usr/lib/python3.8/site-packages/qutebrowser/misc/sql.py", line 205, in _check_ok
    raise_sqlite_error(msg, error)
  File "/usr/lib/python3.8/site-packages/qutebrowser/misc/sql.py", line 129, in raise_sqlite_error
    raise BugError(msg, error)
qutebrowser.misc.sql.BugError: Failed to prepare query "": "Driver not loaded Driver not loaded"
```

That case should probably be handled somehow, and it looks like the cookie filter needs an `@utils.prevent_exceptions` decorator. I'll take a look at that.

The-Compiler commented:
Sorry for the (way) late response from my side as well! :wink: Thanks for the contribution, merged with various (semi-related) changes on top:

- 8e92b29cd Always make sure URLs for per-domain settings are valid
- dad8efa4b Handle invalid URLs in _accept_cookie
- e23d39b37 Ignore exception outside of main thread in crash handler
- 427d8c092 Add @utils.prevent_exceptions() decorator to _accept_cookie
- 4cba2904c Implement per-domain cookie support for QtWebKit
- f8a6d7041 Fix devtools with cookies disabled
- 98fe159f9 Update changelog

@lufte You were totally right in https://github.com/qutebrowser/qutebrowser/pull/4395#issuecomment-434454061 in regards to `RAMCookieJar`, my bad. I ended up adding QtWebKit support in 4cba2904c because it was trivial, and is probably much better than no per-domain support at all.

@Earnestly Thanks again a lot for all the testing and feedback. Should all be fixed on various levels! :+1:
